packages/router/src/router.tsx

```diff
@@ -1,34 +1,116 @@
 // The guts of the router implementation.
 
-import React, { ReactChild, ReactElement } from 'react'
+import React, { ReactChild } from 'react'
 
-import { useAuth as useAuthHook } from '@redwoodjs/auth'
+// TODO: Private routes!
+// import { useAuth as useAuthHook } from '@redwoodjs/auth'
 
 import {
-  Location,
   parseSearch,
   replaceParams,
   matchPath,
-  mapNamedRoutes,
   PageLoader,
   Redirect,
-  LocationContextType,
-  ParamType,
+  useLocation,
+  validatePath,
+  LocationProvider,
 } from './internal'
+import {
+  RouterContextProvider,
+  RouterState,
+  useRouterState,
+} from './router-context'
 import { SplashPage } from './splash-page'
 
+/**
+ * A more specific interface will be generated by
+ * babel-plugin-redwood-routes-auto-loader when a redwood project is built
+ *
+ * Example:
+ * interface AvailableRoutes {
+ *   home: () => "/"
+ *   test: () => "/test"
+ * }
+ */
+interface AvailableRoutes {
+  [key: string]: (args?: Record<string, string>) => string
+}
+
+const namedRoutes: AvailableRoutes = {}
+
 interface RouteProps {
-  page: Spec | React.ComponentType<unknown> | ((props: any) => JSX.Element)
   path?: string
+  page: Spec | React.ComponentType<unknown> | ((props: any) => JSX.Element)
   name?: string
   notfound?: boolean
   redirect?: string
   prerender?: boolean
   whileLoading?: () => ReactChild | null
 }
 
-const Route: React.VFC<RouteProps> = () => {
-  return null
+const Route: React.VFC<RouteProps> = ({
+  path,
+  page,
+  name,
+  redirect,
+  notfound,
+  whileLoading,
+}) => {
+  const location = useLocation()
+  const routerState = useRouterState()
+
+  if (notfound) {
+    // TODO: How should we handle `notfound`?
+    return null
+  }
+
+  if (!path) {
+    throw new Error(`Route "${name}" needs to specify a path`)
+  }
+
+  // Check for issues with the path.
+  validatePath(path)
+
+  // TODO: Figure out how to handle this. This is too late - when first
+  //       rendering a page this hasn't been populated yet and throws errors
+  if (name) {
+    // Create the named route function for this route.
+    namedRoutes[name] = (args = {}) => replaceParams(path, args)
+  }
+
+  const { match, params: pathParams } = matchPath(
+    path,
+    location.pathname,
+    routerState.paramTypes
+  )
+
+  if (!match) {
+    return null
+  }
+
+  const { loading } = routerState.useAuth()
+
+  if (loading) {
+    return <>{whileLoading?.() || null}</>
+  }
+
+  const searchParams = parseSearch(location.search)
+  const allParams = { ...pathParams, ...searchParams }
+
+  if (redirect) {
+    // TODO: Should we use allParams instead of pathParams here?
```

Comment:
I can handle this one myself, just need to do some manual testing to figure out what kind of behavior we want